crates/pipeline_manager/src/api/services/probe.rs

```diff
@@ -0,0 +1,235 @@
+use crate::api::services::kafka::{KafkaServiceProbeRequest, KafkaServiceProbeSuccess};
+use crate::db::DBError;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use thiserror::Error as ThisError;
+use utoipa::ToSchema;
+
+pub trait ServiceProbe {
+    /// Perform the specified probe. It adheres to the timeout, after which
+    /// it gives up on probing the service and returns immediately.
+    fn probe(&self, probe: ServiceProbeRequest, timeout: Duration) -> ServiceProbeResult;
+}
+
+/// Service probe status.
+///
+/// State transition diagram:
+/// ```text
+///    Pending
+///       │
+///       │ (Prober server picks up the probe)
+///       │
+///       ▼
+///   ⌛Running ───► Failure
+///       │
+///       ▼
+///    Success
+/// ```
+#[derive(Deserialize, Serialize, ToSchema, Eq, PartialEq, Debug, Clone, Copy)]
+#[serde(rename_all = "snake_case")]
+pub enum ServiceProbeStatus {
+    /// Probe has not been started.
+    Pending,
+    /// Probe has been picked up by the probe server and is running.
+    Running,
+    /// Probe has finished running and it succeeded.
+    Success,
+    /// Probe has finished running and it failed.
+    Failure,
+}
+
+impl TryFrom<String> for ServiceProbeStatus {
+    type Error = DBError;
+    fn try_from(value: String) -> Result<Self, DBError> {
+        match value.as_str() {
+            "pending" => Ok(Self::Pending),
+            "running" => Ok(Self::Running),
+            "success" => Ok(Self::Success),
+            "failure" => Ok(Self::Failure),
+            _ => Err(DBError::unknown_service_probe_status(value)),
+        }
+    }
+}
+
+impl From<ServiceProbeStatus> for &'static str {
+    fn from(val: ServiceProbeStatus) -> Self {
+        match val {
+            ServiceProbeStatus::Pending => "pending",
+            ServiceProbeStatus::Running => "running",
+            ServiceProbeStatus::Success => "success",
+            ServiceProbeStatus::Failure => "failure",
+        }
+    }
+}
+
+/// Enumeration of all possible service probe requests.
+#[derive(Debug, Serialize, Deserialize, ToSchema, Clone, Eq, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+#[serde(rename_all = "snake_case")]
+pub enum ServiceProbeRequest {
+    /// General test connectivity request.
```

Comment:
Given that this will be part of the external facing documentation, it's good to have a more crisp description of what this tests for.

E.g.: `Tests connectivity to the provided service. For example, for a Kafka service, this probe will check whether the supplied bootstrap configuration and credentials are valid.`

Reply:
Changed to:
```
Tests connectivity to the provided service. For example, for a Kafka service,
this probe will check whether the supplied bootstrap configuration and
credentials are valid by trying to fetch metadata for all topics.
```